File: languages/python/gbdt-par.py
# gbdt-par: parallel multiprocessing variant of gbdt.
# Invocation: python gbdt-par.py <cores> <n>
# argv[1]=cores, argv[2]=n
#
# Parallel decomposition: divide the N samples into `cores` contiguous bands.
# Each worker traverses all B trees for its samples (trees are read-only, static).
# The worker returns a list of (h_band, total_band) where h_band is the partial
# poly-hash for its samples and total_band is the partial sum of acc values.
#
# IMPORTANT: the poly-hash h is order-dependent (h = h*31 + acc+1), so the parent
# must continue hashing from the last h produced by the previous band. Workers
# cannot hash independently and have the parent combine hashes -- the reduction is
# not associative for a poly-hash.
#
# Instead, each worker returns its raw acc values in order, and the parent computes
# the checksum serially over the concatenated acc array (in band=sample order).
# This guarantees core-invariance and matches the serial benchmark exactly.
#
# Core-invariance:
#   - Tree arrays are read-only (feat, thr, leafval built identically).
#   - Each worker computes acc[i] deterministically (same tree traversal as serial).
#   - acc values are assembled in sample order before checksumming.
#   - The final checksum pass is serial and identical to serial gbdt.

import sys
import time
import multiprocessing

P = 1000000007
D = 8
B = 200
F = 8
NODES = (1 << (D + 1)) - 1    # 511
LEAF_START = (1 << D) - 1     # 255


def lcg(s):
    return (s * 1103515245 + 12345) & 0x7FFFFFFF


def build_trees(state):
    """Generate feat, thr, leafval arrays -- identical to serial gbdt.py."""
    feat = [0] * (B * NODES)
    thr = [0] * (B * NODES)
    leafval = [0] * (B * NODES)
    for b in range(B):
        base = b * NODES
        for node in range(LEAF_START):         # internal nodes: feat then thr
            state = lcg(state)
            feat[base + node] = state % F
            state = lcg(state)
            thr[base + node] = state % 256
        for node in range(LEAF_START, NODES):  # leaves
            state = lcg(state)
            leafval[base + node] = state % 10
    return feat, thr, leafval, state


def infer_band(args):
    """Worker: infer all B trees for samples [samp_start, samp_end).

    Returns a list of acc values (one per sample, in order).
    """
    feat, thr, leafval, sample, F_local, samp_start, samp_end = args
    acc_list = []
    for i in range(samp_start, samp_end):
        sbase = i * F_local
        acc = 0
        for b in range(B):
            tbase = b * NODES
            node = 0
            for _ in range(D):
                if sample[sbase + feat[tbase + node]] <= thr[tbase + node]:
                    node = 2 * node + 1
                else:
                    node = 2 * node + 2
            acc += leafval[tbase + node]
        acc_list.append(acc)
    return acc_list


def gbdt_par(cores, n):
    # Build tree arrays (single-threaded, identical to serial).
    feat, thr, leafval, state = build_trees(42)

    # Generate sample data (continues from the tree-building LCG state).
    sample = [0] * (n * F)
    for i in range(n * F):
        state = lcg(state)
        sample[i] = state % 256

    # Band boundaries (floor-division rule).
    bands = []
    for w in range(cores):
        samp_start = w * n // cores
        samp_end = (w + 1) * n // cores
        bands.append((feat, thr, leafval, sample, F, samp_start, samp_end))

    _t0 = time.perf_counter_ns()
    if cores == 1:
        results = [infer_band(bands[0])]
    else:
        with multiprocessing.Pool(processes=cores) as pool:
            results = pool.map(infer_band, bands)
    _t1 = time.perf_counter_ns()
    print(f"COMPUTE_NS {_t1 - _t0}", file=sys.stderr)

    # Assemble acc values in sample order (deterministic).
    # Compute checksum serially -- identical to serial gbdt.
    h = 0
    total = 0
    for band_acc in results:
        for acc in band_acc:
            h = (h * 31 + acc + 1) % P
            total = (total + acc) % P

    return h, total


if __name__ == "__main__":
    cores = int(sys.argv[1]) if len(sys.argv) > 1 else 1
    n = int(sys.argv[2]) if len(sys.argv) > 2 else 5000
    h, sec = gbdt_par(cores, n)
    print(h)
    print("gbdt(%d) = %d" % (n, sec))

File: languages/ruby/gbdt-par.rb
# gbdt-par: parallel PROCESS variant of gbdt (scaling track).
# Invocation: ruby gbdt-par.rb <cores> <n>   (ARGV[0]=cores, ARGV[1]=n)
#
# MRI has a GVL, so the fair CPU-parallel primitive is PROCESSES via Process.fork.
# Decompose the N samples into `cores` contiguous bands. Each forked worker traverses
# all B trees for its samples (the tree arrays are read-only / static) and returns its
# per-sample acc values, in sample order, to the parent through a pipe. The parent
# concatenates the bands in index order and runs the serial poly-hash checksum loop.
#
# IMPORTANT: the poly-hash h is order-dependent (h = h*31 + acc+1), so it is NOT an
# associative reduction - workers cannot hash their bands independently. Instead each
# worker returns its raw acc values and the parent hashes the concatenated array
# serially, exactly as the serial benchmark does.
#
# Core-invariance:
#   - Tree arrays (feat, thr, leafval) are built identically and never written.
#   - Each worker computes acc[i] with the same tree traversal as serial.
#   - acc values are assembled in sample order before checksumming.
#   - The final checksum pass (h and total) is serial and identical to serial gbdt.

P          = 1000000007
D          = 8
B          = 200
F          = 8
NODES      = 511  # 2^(D+1) - 1
LEAF_START = 255  # 2^D - 1

# Infer all B trees for samples [samp_start, samp_end); return acc values in order.
def infer_band(feat, thr, leafval, sample, samp_start, samp_end)
  acc_list = Array.new(samp_end - samp_start, 0)
  i = samp_start
  while i < samp_end
    sbase = i * F
    acc = 0
    b = 0
    while b < B
      tbase = b * NODES
      node = 0
      D.times do
        if sample[sbase + feat[tbase + node]] <= thr[tbase + node]
          node = 2 * node + 1
        else
          node = 2 * node + 2
        end
      end
      acc += leafval[tbase + node]
      b += 1
    end
    acc_list[i - samp_start] = acc
    i += 1
  end
  acc_list
end

# Run inference over `cores` forked workers; return per-band acc lists in band order.
# cores==1 runs inline (no fork).
def run_infer(cores, feat, thr, leafval, sample, bands)
  return bands.map { |ss, se| infer_band(feat, thr, leafval, sample, ss, se) } if cores == 1

  readers = []
  bands.each do |ss, se|
    r, w = IO.pipe
    pid = Process.fork do
      r.close
      Marshal.dump(infer_band(feat, thr, leafval, sample, ss, se), w)
      w.close
      exit!(0)
    end
    w.close
    readers << [r, pid]
  end

  readers.map do |r, pid|
    data = r.read
    r.close
    Process.waitpid(pid)
    Marshal.load(data)
  end
end

def gbdt_par(cores, n)
  feat    = Array.new(B * NODES, 0)
  thr     = Array.new(B * NODES, 0)
  leafval = Array.new(B * NODES, 0)

  state = 42
  B.times do |b|
    base = b * NODES
    LEAF_START.times do |node|
      state = (state * 1103515245 + 12345) & 0x7fffffff
      feat[base + node] = state % F
      state = (state * 1103515245 + 12345) & 0x7fffffff
      thr[base + node]  = state % 256
    end
    (LEAF_START...NODES).each do |node|
      state = (state * 1103515245 + 12345) & 0x7fffffff
      leafval[base + node] = state % 10
    end
  end

  sample = Array.new(n * F, 0)
  (n * F).times do |i|
    state = (state * 1103515245 + 12345) & 0x7fffffff
    sample[i] = state % 256
  end

  bands = (0...cores).map do |w|
    [w * n / cores, (w + 1) * n / cores]
  end

  # COMPUTE_NS region (to stderr): the parallel inference - fork+IPC of the B-tree
  # traversal over all samples, results returned in band order. Excludes the tree build
  # + sample data-gen above and the serial checksum below, so the scaling track's
  # speedup excludes interpreter startup + data-gen.
  t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond)
  band_results = run_infer(cores, feat, thr, leafval, sample, bands)
  t1 = Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond)
  $stderr.puts("COMPUTE_NS #{t1 - t0}")

  # Serial checksum over the acc values, in sample order - identical to serial gbdt.
  h     = 0
  total = 0
  band_results.each do |band_acc|
    band_acc.each do |acc|
      h     = (h * 31 + acc + 1) % P
      total = (total + acc)       % P
    end
  end
  [h, total]
end

cores = ARGV[0] ? ARGV[0].to_i : 1
n = ARGV[1] ? ARGV[1].to_i : 5000
h, total = gbdt_par(cores, n)
puts h
puts "gbdt(#{n}) = #{total}"
